Generate OSGi metadata into apfloat jars to make it OSGi compliant

HannesWell opened the pull request:
This allows usage of the following modules in an OSGi runtime out of the box:
- `apfloat`
- `apfloat-aparapi`
- `apfloat-jscience`

All OSGi metadata are generated at build-time using the [bnd-maven-plugin](https://github.com/bndtools/bnd/tree/master/maven-plugins/bnd-maven-plugin) provided by [bndtools](https://bnd.bndtools.org/).

Additionally let the bnd-maven-plugin generate the services file for the apfloat module together with corresponding OSGi metadata to make service usage work in OSGi runtimes. In order to make this work add a _compile-time only_ dependency to ['biz.aQute.bnd:biz.aQute.bnd.annotation:7.0.0'](https://mvnrepository.com/artifact/biz.aQute.bnd/biz.aQute.bnd.annotation/7.0.0).

In case you are interested in having the module-info.java for the Java Platform Module System generated as well, this could also be generated using the bnd-maven-plugin:
https://bnd.bndtools.org/chapters/330-jpms.html

Last but not least, this PR fixes and shortens the build instructions in the Readme, `jarsigner.skip` is the property to skip signing and not `gpg.skip`.

mtommila commented:
Hello,

I'm not really familiar with OSGi runtimes so I have some questions about this.

* I suppose also apfloat-applet and apfloat-calc should have the same handling as they could reasonably be used by other projects
* Although it seems like the default package in apfloat-applet would not work
* I'm not sure about the use case for the `@ServiceProvider` annotation, to me it seems unlikely that anything else than the apfloat library itself would use these service providers, and the apfloat library isn't reading that annotation (it uses a system property or a value from a Java resource bundle, or a hard-coded default value to determine the BuilderFactory implementation class; the sample applications use `java.util.ServiceLoader`)
* Minor detail but I want the ServiceProvider entries in a specific order (this affects the order in which they appear in the sample applications UI)
* Is the actual use case of this PR to have the OSGi metadata in the manifest files, or does it actually matter which mechanism is used to generate the metadata
* Only the `apfloat` module is multi-release so the other modules don't need that in the manifest, but this is also a minor detail
* Why would bnd-maven-plugin be better than some other tools for doing this (e.g. maven-bundle-plugin), or would it be sufficient to just have the OSGi metadata there in whatever way (is it e.g. reasonable to do manually without any Maven plugin)
* How do I test this

In general, I would be happy to have the OSGi metadata in the apfloat artifacts, if it doesn't interfere with anything else and if I can choose the most elegant mechanism how to do that

For example: [diff.patch](https://github.com/mtommila/apfloat/files/13462592/diff.patch)

HannesWell commented:
> Hello,
> 
> I'm not really familiar with OSGi runtimes so I have some questions about this.

No problem, I'll try to answer them. :)
I also want to mention that I don't use apfloat directly, but use [Symja/matheclipse](https://github.com/axkr/symja_android_library/), which depends on apfloat. Therefore I'm unfortunately not very familiar with this library.

> 
>     * I suppose also apfloat-applet and apfloat-calc should have the same handling as they could reasonably be used by other projects
> 
>     * Although it seems like the default package in apfloat-applet would not work

I tried that initially as well as stumbled over the exact same problem and after looking into the description of two mentioned modules I assumed that it is very unlikely that they are used in an OSGi runtime, since (as far as I understood it) they are standalone applications or intended for applets. Therefore I tried to go the simple way and just skipped them.

> 
>     * I'm not sure about the use case for the `@ServiceProvider` annotation, to me it seems unlikely that anything else than the apfloat library itself would use these service providers, and the apfloat library isn't reading that annotation (it uses a system property or a value from a Java resource bundle, or a hard-coded default value to determine the BuilderFactory implementation class; the sample applications use `java.util.ServiceLoader`)
> 
Yes this annotation is only processed by bnd-tools/bnd-maven-plugin to generate the services file and the corresponding OSGi metadata to support loading these services in OSGi runtimes from other 'bundles/modules'. They have a retention policy `CLASS` and are therefore not available at runtime and the dependency scope is provided to make it a compile-time only dependency. Unfortunately there is no such scope named like this, but Maven suggests to use provided in that case:
https://maven.apache.org/general.html#scope-provided 

This is a bit complicated topic, but to summaries it, because in OSGi each bundle/module has its own classloader loading services from other bundles is not possible unless one depends on them, which contradict the intended inversion-of-control and thus would make it useless. Therefore the services have to be 'registered/declared' in the MANIFEST.MF so that. If you are interested in more details, there is a specification about that topic or I can answer further questions:
https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.loader.html

But if that is all necessary depends on if these factories need to be loaded as services by other bundles at any time or if this was just used in the samples for some specific reason?

>     * Minor detail but I want the ServiceProvider entries in a specific order (this affects the order in which they appear in the sample applications UI)

OK, then the file needs to be left as it is. But that's not a problem. The OSGi metadata necessary for service loading, can also be added explicitly (if they are necessary at all).


>     * Is the actual use case of this PR to have the OSGi metadata in the manifest files, or does it actually matter which mechanism is used to generate the metadata

The main purpose of this PR is to add OSGi metadata.

> 
>     * Only the `apfloat` module is multi-release so the other modules don't need that in the manifest, but this is also a minor detail

Yes I noticed that too, but checked the result and they look like it didn't harm (although it was not necessary), but it can be done right with more specialized configuration.

>     * Why would bnd-maven-plugin be better than some other tools for doing this (e.g. maven-bundle-plugin), or would it be sufficient to just have the OSGi metadata there in whatever way (is it e.g. reasonable to do manually without any Maven plugin)
> 

The `maven-bundle-plugin` is based on BND: https://felix.apache.org/documentation/_attachments/components/bundle-plugin/
Therefore I would suggest to use the maven-plugin provided by BND instead since it is more up to date.
In general the maven-bundle-plugin should also work, but IIRC Multi-Release jar support was not working well before or not at all before BND 7.0.0 and there is no version of the maven-bundle-plugin yet that picks that update up.

In general it would also be possible to use a hand-crafted `MANIFEST.MF` and add the entries explicitly using the maven-jar-plugin `archive->manifestEntries`. But this would require more release engineering work to always keep the bundle and package versions in sync or consideration and adjustments if dependencies change.
In general generating the `MANIFEST.MF` entries is usually the safest way and the way of least work. The generated metadata are once setup usually very good/correct.


>     * How do I test this
> 
This depends on your development environment. Are you using Eclipse?

> 
> In general, I would be happy to have the OSGi metadata in the apfloat artifacts, if it doesn't interfere with anything else and if I can choose the most elegant mechanism how to do that

Glad to hear that and no it shouldn't interfere with anything else.

> 
> For example: [diff.patch](https://github.com/mtommila/apfloat/files/13462592/diff.patch)

Yes that looks good.
I updated this PR to use that suggestion and added the OSGi metadata for the services of the apfloat module (assuming that they are intended to be used by the outside world).

mtommila commented:
Thank you for the helpful information. Some more comments:

The calculator:

It can be used so that it inputs data from a Reader, and writes the result to a Writer, so in principle it's possible to embed it anywhere that reads and writes text. Yes there is a command-line application that uses stdin and stdout for the input and output, and also an Applet and an AWT application (in the samples) that just redirect the input and output to the GUI. But it's layered so that it's also possible to use it anywhere else, as a "library".

The apfloat-applet library:

The main concern that anybody would have about this is that the default configuration of the apfloat library is that numbers that are larger than a threshold (which by default depends on the max heap of the JVM) are stored on disk. I'm guessing that this is undesirable (or outright not permitted) in many applications, for example in unsigned applets (well, nobody uses applets anymore). Anyways, if the apfloat-applet library is in the classpath, then it changes the default behavior so that it never stores numbers on disk, no matter how big. It only stores numbers in memory. But then this has the drawback that you can't make calculations with maximum precision (where just one number is in memory at a time and otherwise everything is stored on disk). On modern computers this "maximum precision" is in the order of billions of digits of precision so this problem only applies to some use cases (like calculating pi to 10 billion digits).

Service loading:

I really think that the the OSGi metadata for service loading is not necessary. First of all, the apfloat "SPI" was always intended to be used only internally by the apfloat library itself. The SPI has four different implementations in the library itself. Why it exists is that on different CPUs the different implementations can perform very differently - there exists no single implementation that would perform best on all possible CPUs. First of all 32-bit and 64-bit computers have a big performance difference in using int or long (int is faster on 32-bit and long is faster on 64-bit), and in specific cases, some CPUs have poor integer (especially multiplication) performance but fast floating-point performance, or vice versa.

If anything, I always thought that I could add more implementations of the BuilderFactory, like the aparapi implementations were added later. I never thought that any third party would use the SPI implementations, it was never intended to be used that way. The SPI is not supposed to be public. Also I may make backwards-incompatible changes to the SPI in the future (like I have made in the past) so I don't think it should be ever public, to avoid trouble.

As said the library itself does not use java.util.ServiceLoader so I think there will be no problem with the OSGi classloading. The library just instantiates the class directly, by class name. The java.util.ServiceLoader is only used by the sample applications.

I tried to make this layering (high level / low level implementation) as elegant as possible, but in retrospect it's perhaps a bit over-engineered.

The apfloat-aparapi library is actually just an implementation of the SPI, so I wonder how it would work with OSGi? If so configured, the apfloat library would instantiate a class from the apfloat-aparapi library (but directly with Java reflection, not using java.util.ServiceLoader).

The Maven plugin:

I guess the maven-bundle-plugin does not work with newer than Java 11 so it seems inferior. Generating the manifests by hand would likely be a gigantic hassle so using the bnd plugin is probably then the best choice.

Eclipse:

Yes I'm using Eclipse.

HannesWell commented:
Thanks for the elaboration.

> If anything, I always thought that I could add more implementations of the BuilderFactory, like the aparapi implementations were added later. I never thought that any third party would use the SPI implementations, it was never intended to be used that way. The SPI is not supposed to be public. Also I may make backwards-incompatible changes to the SPI in the future (like I have made in the past) so I don't think it should be ever public, to avoid trouble.

Understand, then I agree that these entries are not necessary. Thanks for the elaboration.
For completeness I want to stress that the extra entries that were present are intended to make the services of apfloat available as a provider. If apfloat would use `ServiceLoader.load()`, it would be a consumer and therefore would require different entries as well. But from all you have described, it should not be necessary.

Thus I have removed the corresponding entries again.

> The apfloat-aparapi library is actually just an implementation of the SPI, so I wonder how it would work with OSGi? If so configured, the apfloat library would instantiate a class from the apfloat-aparapi library (but directly with Java reflection, not using java.util.ServiceLoader).

This will probably not work in OSGi, because `apfloat` does not depend on `apfloat-aparapi`, it is only the other way round. Therefore the `apfloat` bundle cannot access/load classes from `apfloat-aparapi`, regardless if this is done through reflection or usual code.
In that case you either would have to use [DynamicImport-Package](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.module.html#framework.module.dynamicimportpackage) or the full Service-Loader mechanism (but in this case I would go simply go for the former approach).


> The Maven plugin:
> 
> I guess the maven-bundle-plugin does not work with newer than Java 11 so it seems inferior. Generating the manifests by hand would likely be a gigantic hassle so using the bnd plugin is probably then the best choice.

Fully agree.


> * How do I test this

In Eclipse you can for example create a new Plug-in project that uses apfloat and then define and launch an application that starts this Plugin.

Use the `New Project...` wizard to create a new `Plug-in Project` and make sure that on the second page `Generate an activator` is checked. In the generated Plug-in project go to the MANIFEST.MF and add in the `Dependencies` tab a dependency to the `org.apfloat` bundle or to one of its packages.
In the generated Activator (the only class in the project) add in the start() method for example:
```
Apfloat a = new Apfloat(10);
System.out.println(a.multiply(a));
```
Then you can go into the Run/Debug configuration dialog and create a new `Eclipse Application` launch. Then in the Plug-ins tab select `Launch with: Features selected below` and select for example the `org.eclipse.platform` feature.
In order to include the created Plugin/Bundle click on the right side, `Select Plug-ins...` and add the just created Plug-in.
For that plugin set in the Auto-Start column (the right most) the value `true`.
Then run the so configured Eclipse application and the activator method should be executed during its start-up.

Using apfloat in a Plugin/Bundle will only work as long as you have the apfloat modules/projects imported in your workspace and the changes of this PR applied. The Maven integration of Eclipse (m2e) will then generate the OSGi metadata as part of the workspace build and will make it available to the Eclipse Plugin Development Environment (PDE).

mtommila commented:
Regarding apfloat-aparapi, would it help anything if the apfloat module had an "optional" Maven dependency to apfloat-aparapi? Would this allow apfloat to load classes from apfloat-aparapi?

mtommila commented:
Hmm, obviously that won't work as long as it creates a cyclic dependency between apfloat and apfloat-aparapi.

mtommila commented:
Also, if I only specify `-Djarsigner.skip` (and not `-Dgpg.skip`) then GPG still does prompt me for the password for signing.

The instructions could be shortened a bit by omitting the `=true` though, i.e. to have just `-Dgpg.skip -DskipTests`

Isn't it so that the jarsigner and gpg plugin do two different things? The jarsigner is used only for creating a signed applet jar. However the gpg plugin is needed (only) when making a release to the Sonatype OSS central repository, and it's needed for all artifacts, not just the signed applet jar in the apfloat-samples?

HannesWell commented:
> Hmm, obviously that won't work as long as it creates a cyclic dependency between apfloat and apfloat-aparapi.

Yes. I think just adding `DynamicImport-Package: org.apfloat.aparapi` to the bnd-instructions of the `apfloat` module should allow `apfloat` to load the aperapi classes through reflection.
But if I understand it right, that's currently not done, is it?


> Isn't it so that the jarsigner and gpg plugin do two different things? The jarsigner is used only for creating a signed applet jar. However the gpg plugin is needed (only) when making a release to the Sonatype OSS central repository, and it's needed for all artifacts, not just the signed applet jar in the apfloat-samples?

Ups. Yes that's totally right. I think both should be skipped the gpg-plugin and the jarsigner-plugin. I just didn't notice the gpg is active since I have gpg signing configured in my local maven config.
I updated the README change. I hope it is fine now.

mtommila commented:
Since the readme file says "To build the library quickly..." and the jarsigner only takes a fraction of a second to run, it's probably faster to not type `-Djarsigner.skip` and instead just let the jarsigner run. That is, if you want to build the library as fast as possible (including the time to type the command to build the library).

(Compared to running the unit tests which takes around 10 minutes, or running the gpg plugin which prompts you for the password and if you don't give it, it fails after a timeout)

Anyways, yes the apfloat module would load the apfloat-aparapi classes through reflection. Would this `DynamicImport-Package: org.apfloat.aparapi` just be added to the bnd plugin configuration in the apfloat module?

HannesWell commented:
> Since the readme file says "To build the library quickly..." and the jarsigner only takes a fraction of a second to run, it's probably faster to not type `-Djarsigner.skip` and instead just let the jarsigner run. That is, if you want to build the library as fast as possible (including the time to type the command to build the library).

Yes, I agree with you that performance wise it is not relevant, but at least for me the build failed when the jarsigner-plugin was executed. But I haven't investigated it further if this is just due to my local maven configuration.
If you prefer to keep this as it is I can also revert that first commit that changes the readme (since it is actually not related to the main intended of this PR it wasn't the cleanest workflow to add it here anyways).

> 
> Anyways, yes the apfloat module would load the apfloat-aparapi classes through reflection. Would this `DynamicImport-Package: org.apfloat.aparapi` just be added to the bnd plugin configuration in the apfloat module?

Yes exactly, that should be it. You can just put it in a new line below `Multi-Release: true`.

mtommila commented:
Indeed, you are right, the jarsigner would fail if the jar signing key doesn't exist (I forgot that since I've had the key for years). This is explained in the readme file but it's obviously far easier to skip the jar signing than to first generate the signing key and then sign the jar (especially as running a signed applet is nowadays exceedingly difficult so there's no reason why anybody would ever do that).

So it would make more sense if the readme file read something like:

To build the library quickly, without running unit tests (takes about 10 minutes) and without signing with GPG run:

`mvn clean install -Dgpg.skip -Djarsigner.skip -DskipTests`

To build the signed applet files, you need to first generate a signing key, e.g. with:

`keytool -genkeypair -validity 21915 -dname "cn=Your Name, o=example.com" -storepass password -keypass password -alias mykey`

and then build without `-Djarsigner.skip`.

HannesWell commented:
> So it would make more sense if the readme file read something like:
> 
> To build the library quickly, without running unit tests (takes about 10 minutes) and without signing with GPG run:
> 
> `mvn clean install -Dgpg.skip -Djarsigner.skip -DskipTests`
> 
> To build the signed applet files, you need to first generate a signing key, e.g. with:
> 
> `keytool -genkeypair -validity 21915 -dname "cn=Your Name, o=example.com" -storepass password -keypass password -alias mykey`
> 
> and then build without `-Djarsigner.skip`.

Agree. Updated the first commit accordingly.

HannesWell commented:
> I pushed a commit,

Thanks. Should I integrate those changes into the two previous commits or do you want to keep it separated?

> but when I try to test this and create a new Plugin project in Eclipse (based on the Hello World Command template) it gives this error:
> 
> Cannot invoke "org.eclipse.pde.core.plugin.IPluginModelBase.getPluginBase()" because "model" is null

You don't need to select one of the templates. Nevertheless I tried to reproduce this issue, but for me it worked.
Are you using the latest release of Eclipse (Wednesday next week 2023-12 will be released)?
If it still does not work can you please report this to https://github.com/eclipse-pde/eclipse.pde/issues with the detailed steps to reproduce and the stack-trace from the `Error Log` view?

mtommila commented:
Please do squash/integrate that commit with the two previous commits.

I'm a bit unsure how the dependencies (at least the classes org.eclipse.ui.plugin.AbstractUIPlugin and org.osgi.framework.BundleContext) are supposed to be added to the classpath? Which version of Java should I use?

mtommila commented:
I'm unsure that should I conclude that it's not actually working, or do I have something configured incorrectly in my test Eclipse plugin project.

I have added the apfloat library to the plugin project's classpath, ticked the checkbox that it's exported, the org.apfloat module is listed in the dependencies, in the imported packages when I open the pde.bnd file (although: it has a red "X" marker but no other information what error is there) and it appears in the plugin project's generated MANIFEST.MF file in the Import-Package section.

Then, when I try to run the Eclipse application, it gives the error `Missing Constraint: Import-Package: org.apfloat; version="0.0.0"`

But looking at the apfloat library that is in the plugin project's classpath, it does have `Export-Package: org.apfloat` etc. so it would seem to be correct (?).

HannesWell commented:
> Please do squash/integrate that commit with the two previous commits.
> 
Just did that. Sorry for the delayed reply.

> I'm a bit unsure how the dependencies (at least the classes org.eclipse.ui.plugin.AbstractUIPlugin and org.osgi.framework.BundleContext) are supposed to be added to the classpath? Which version of Java should I use?

These dependencies should come from the so called _Target-Platform_. In the Eclipse Plug-in Developers Environment (PDE) this is the set of bundles against which the Plugin projects in the workspace are compiled.
You can fully configure that with a so called target file, but by default the running Eclipse is just used as Target-Platform and for this case it should be sufficient.


> I'm unsure that should I conclude that it's not actually working, or do I have something configured incorrectly in my test Eclipse plugin project.

First of all you need to ensure that you have PDE installed (i.e. the Feature `org.eclipse.pde`). You could install it via `Help -> Install New Software...` or if you don't want to change your existing installation you can download the 

[Eclipse IDE for Eclipse Committers](https://www.eclipse.org/downloads/packages/release/2023-09/r/eclipse-ide-eclipse-committers) or the [Eclipse Modeling Tools](https://www.eclipse.org/downloads/packages/release/2023-09/r/eclipse-modeling-tools) pre-build package (or use the Oomph-Installer to install the corresponding Product).
Sorry I didn't mention that before because I (falsely) assumed that PDE would be installed all the time.

> 
> I have added the apfloat library to the plugin project's classpath, ticked the checkbox that it's exported, the org.apfloat module is listed in the dependencies, in the imported packages when I open the pde.bnd file (although: it has a red "X" marker but no other information what error is there) and it appears in the plugin project's generated MANIFEST.MF file in the Import-Package section.
> 

It looks like you have created a project with automated Manifest generation. This is a brand new way to define Plugin projects in PDE, but it is harder to learn than the classic PDE Plugin projects.
In general was more thinking about creating a classic PDE Plugin project as it is described here:
https://www.vogella.com/tutorials/EclipsePlugin/article.html#exercise-creating-and-using-a-first-plug-in
But you can skip the last part where a template is selected.

But give me a second, I will assemble a simple project for you plus a launch configuration.

HannesWell commented:
> Which version of Java should I use?

Eclipse requires Java-17 so you can use just that, but Java-21 or in-between should also work.

Besides the above mentioned PDE, you also need a Plugin from Eclipse Maven integration that bridges to PDE, which is named `org.eclipse.m2e.pde.feature`.

If you use the [Eclipse IDE for Eclipse Committers](https://www.eclipse.org/downloads/packages/release/2023-09/r/eclipse-ide-eclipse-committers), both is already installed:
- `org.eclipse.pde` with name `M2E - PDE Integration`
- `org.eclipse.m2e.pde.feature` with name `Eclipse Plug-in Development Environment`

In case you want to install both, they are probably easier to find through their name.

Please find attached a minimal example project:
[apfloat.osgi.playground.zip](https://github.com/mtommila/apfloat/files/13518817/apfloat.osgi.playground.zip)

If you import it into your workspace, where apfloat is checked-out at this PR's branch and with the requirements mentioned above everything should compile without errors.

The attached project is a PDE Plugin project that has a `META-INF/MANIFEST.MF` file to control its dependencies at compile and runtime.
On the other hand apfloat is a Maven project, which is why org.eclipse.m2e.pde.feature is necessary to tell PDE how it should handle the Maven project `aploat`.


The attached project also contains the `product.with.apfloat.launch`. If you right-click on it and select `Run As` or `Debug As` and then `product.with.apfloat` it launches a very basic Eclipse application (which is an OSGi application) that calls during its start-up the `apfloat.osgi.playground.Activator.start()` method.
You can of course change the code. If you want to add other dependencies you can add them in the Dependencies tab of the `META-INF/MANIFEST.MF` file, either as required bundle (to get all packages of that bundle/plugin) or via Import-Package (to only get access to that package). With these kind of projects you should never have the need to adjust the classpath directly.

I hope this helps.

mtommila commented:
Yes, it seems to work with this still added to the `apfloat-dependencies.target`:
```
				<dependency>
					<groupId>com.aparapi</groupId>
					<artifactId>aparapi</artifactId>
					<version>3.0.0</version>
					<type>jar</type>
				</dependency>
```

HannesWell commented:
> Yes, it seems to work with this still added to the `apfloat-dependencies.target`:
> 
> ```
> 				<dependency>
> 					<groupId>com.aparapi</groupId>
> 					<artifactId>aparapi</artifactId>
> 					<version>3.0.0</version>
> 					<type>jar</type>
> 				</dependency>
> ```

Oh, yes I forgot that, but I'm glad that it is now working. As you noticed the `Maven` target-locations use the usual Maven-Dependencies syntax.

So do you think this is now ready or is there something I have to add?

mtommila commented:
Alas, I have no plan for the next release date. There are a few tickets open that suggest various new functions being implemented, so it might take a while.

In the meantime, I have deployed a snapshot to https://oss.sonatype.org/content/repositories/snapshots/ if that helps anything.

HannesWell commented:
Thank you for the update.

> In the meantime, I have deployed a snapshot to https://oss.sonatype.org/content/repositories/snapshots/ if that helps anything.

Unfortunately I can't use that, mainly for technical reasons.
But I can just wait for the next release.

mtommila commented:
The new apfloat version 1.13.0 has been released.

HannesWell commented:
Awesome! Thank you.
